### ISSUE TITLE
Uncaught exception when composer.json references packages.drupal.org

### ISSUE THREAD

**Issue by mscottford** [ecoAPM/php-libyear#1]
When running `php-libyear` against a `composer.json` file that references a package which is not hosted at packagist.org, but instead is hosted in another composer repository, an uncaught exception is encountered, because the pakagist.org API is returning a 404 when the package's information is asked for. The full error output can be found below, along with a minimal composer.json file (and corresponding lock file) which triggers the error.

I suspect that this same error would be triggered by a `composer.json` file which is referencing asset-packagist.org or has a "vcs" or "git" repository configured. Both of those are also contained in the composer.json file where this issue was discovered.

composer.json:
```
{
	"name": "packages/drupal_example",
	"description": "A minimal example of a composer manifest that uses the drupal package database",
	"repositories": [
		{
			"type": "composer",
			"url": "https://packages.drupal.org/7"
		}
	],
	"require": {
		"php": ">=7.1",
		"drupal/composer_autoloader": "^1.0"
	}
}
```

composer.lock:
```
{
    "_readme": [
        "This file locks the dependencies of your project to a known state",
        "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
        "This file is @generated automatically"
    ],
    "content-hash": "22ef2fdbe0447eeba8c946fec05ea678",
    "packages": [
        {
            "name": "drupal/composer_autoloader",
            "version": "1.3.0",
            "source": {
                "type": "git",
                "url": "https://git.drupalcode.org/project/composer_autoloader.git",
                "reference": "7.x-1.3"
            },
            "dist": {
                "type": "zip",
                "url": "https://ftp.drupal.org/files/projects/composer_autoloader-7.x-1.3.zip",
                "reference": "7.x-1.3",
                "shasum": "9e5626ade0134d7c2375f724fc359c09648c03c9"
            },
            "require": {
                "drupal/drupal": "~7.0"
            },
            "conflict": {
                "drupal/composer_autoload": "*",
                "drupal/composer_manager": "*"
            },
            "type": "drupal-module",
            "extra": {
                "branch-alias": {
                    "dev-1.x": "1.x-dev"
                },
                "drupal": {
                    "version": "7.x-1.3",
                    "datestamp": "1520741884",
                    "security-coverage": {
                        "status": "covered",
                        "message": "Covered by Drupal's security advisory policy"
                    }
                }
            },
            "notification-url": "https://packages.drupal.org/7/downloads",
            "license": [
                "GPL-2.0+"
            ],
            "authors": [
                {
                    "name": "davidwbarratt",
                    "homepage": "https://www.drupal.org/user/411675"
                }
            ],
            "description": "Simple module to load the Composer autoload.php file.",
            "homepage": "https://www.drupal.org/project/composer_autoloader",
            "support": {
                "source": "https://git.drupalcode.org/project/composer_autoloader",
                "issues": "http://drupal.org/project/issues/composer_autoloader"
            }
        },
        {
            "name": "drupal/drupal",
            "version": "7.69.0",
            "source": {
                "type": "git",
                "url": "https://git.drupalcode.org/project/drupal.git",
                "reference": "7.69"
            },
            "dist": {
                "type": "zip",
                "url": "https://ftp.drupal.org/files/projects/drupal-7.69.zip",
                "reference": "7.69",
                "shasum": "424bbf205fc6f8b28b348d42206b395fca26724f"
            },
            "type": "drupal-core",
            "extra": {
                "branch-alias": {
                    "dev-7.x": "7.x-dev"
                },
                "drupal": {
                    "version": "7.69",
                    "datestamp": "1576698943",
                    "security-coverage": {
                        "status": "covered",
                        "message": "Covered by Drupal's security advisory policy"
                    }
                }
            },
            "notification-url": "https://packages.drupal.org/7/downloads",
            "license": [
                "GPL-2.0-or-later"
            ],
            "authors": [
                {
                    "name": "Dries",
                    "homepage": "https://www.drupal.org/user/1"
                },
                {
                    "name": "Drupal",
                    "homepage": "https://www.drupal.org/user/3"
                },
                {
                    "name": "Fabianx",
                    "homepage": "https://www.drupal.org/user/693738"
                },
                {
                    "name": "Gábor Hojtsy",
                    "homepage": "https://www.drupal.org/user/4166"
                },
                {
                    "name": "alexpott",
                    "homepage": "https://www.drupal.org/user/157725"
                },
                {
                    "name": "catch",
                    "homepage": "https://www.drupal.org/user/35733"
                },
                {
                    "name": "cilefen",
                    "homepage": "https://www.drupal.org/user/1850070"
                },
                {
                    "name": "drumm",
                    "homepage": "https://www.drupal.org/user/3064"
                },
                {
                    "name": "effulgentsia",
                    "homepage": "https://www.drupal.org/user/78040"
                },
                {
                    "name": "larowlan",
                    "homepage": "https://www.drupal.org/user/395439"
                },
                {
                    "name": "lauriii",
                    "homepage": "https://www.drupal.org/user/1078742"
                },
                {
                    "name": "mcdruid",
                    "homepage": "https://www.drupal.org/user/255969"
                },
                {
                    "name": "plach",
                    "homepage": "https://www.drupal.org/user/183211"
                },
                {
                    "name": "webchick",
                    "homepage": "https://www.drupal.org/user/24967"
                },
                {
                    "name": "xjm",
                    "homepage": "https://www.drupal.org/user/65776"
                },
                {
                    "name": "yoroy",
                    "homepage": "https://www.drupal.org/user/41502"
                }
            ],
            "homepage": "https://www.drupal.org/project/drupal",
            "support": {
                "source": "https://git.drupalcode.org/project/drupal"
            }
        }
    ],
    "packages-dev": [],
    "aliases": [],
    "minimum-stability": "stable",
    "stability-flags": [],
    "prefer-stable": false,
    "prefer-lowest": false,
    "platform": {
        "php": ">=7.1"
    },
    "platform-dev": []
}
```

Output:

```
bash-5.0# php libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar libmetrics/LibMetrics.Test/fixtures/php/drupal/
Gathering information...

Fatal error: Uncaught GuzzleHttp\Exception\ClientException: Client error: `GET https://repo.packagist.org/packages/drupal/composer_autoloader.json` resulted in a `404 Not Found` response:
{"status":"error","message":"Package not found"}
 in phar:///app/libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113
Stack trace:
#0 phar:///app/libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar/vendor/guzzlehttp/guzzle/src/Middleware.php(65): GuzzleHttp\Exception\RequestException::create(Object(GuzzleHttp\Psr7\Request), Object(GuzzleHttp\Psr7\Response))
#1 phar:///app/libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar/vendor/guzzlehttp/promises/src/Promise.php(203): GuzzleHttp\Middleware::GuzzleHttp\{closure}(Object(GuzzleHttp\Psr7\Response))
#2 phar:///app/libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar/vendor/guzzlehttp/promises/src/Promise.php(156): Guzz in phar:///app/libmetrics/LibMetrics.Test/fixtures/php/large/tmp/php-libyear/dist/libyear.phar/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php on line 113
bash-5.0# exit
```

**Comment by mscottford** [comment]
I'm working on addressing this issue for github.com/corgibytes/libmetrics, because it is showing up there, too. I'll drop notes about my learnings here.

**Comment by mscottford** [comment]
URLs of the form `https://repo.packagist.org/packages/drupal/composer_autoloader.json` are only going to work for packages hosted on `packagist.org`. For others, you'll need to lookup `packages.json` and parse it the way that `composer` does. At `https://packages.drupal.com/7/packages.json`, the `providers-url` and `provider-includes` fields. Documentation for those can help with implementation details: https://getcomposer.org/doc/05-repositories.md#provider-includes-and-providers-url.

The documentation for `packages.json` claims that a `packages` field is required (https://getcomposer.org/doc/05-repositories.md#package), but that must be a mistake, because the drupal one does not include that field, and `composer` is able to pull packages from it just fine.

**Comment by mscottford** [comment]
`https://asset-packagist.org/packages.json` does not include a `packages` field, either. So I'm guessing that doc is out of date.

**Comment by mscottford** [comment]
So to find the version information for `drupal/composer_autoloader`, I have to visit each entry in `provider-includes` from `packages.json`. The resulting file will include a `providers` field, which contains a field name for each package that is in that list. If the package that's being looked for is not there, then you have to move on to the next `provider-includes` entry. When looking for `drupal/composer_autoloader`, I eventually find it in: https://packages.drupal.org/7/drupal/provider-2018-1$8f0bd09fb74e65db3dc030e24608443c6187d5709e22caca2edf24ef453af7ca.json.

This file only provides me with a hash that I can use to construct a valid URL to pull up the version information. The `packages.json` file contains the URL template that should be used for this in the `providers-url` field. I take the package name and the hash from the provider-include file and pulg them into the template to get https://packages.drupal.org/7/drupal/composer_autoloader$6a9755ef98f1299fa64e6bab7f23a515944324c012e77391f1b3882dca98c18c.json. And that file contains all of the information that we need.

I suspect that this same approach can be used for packagist.org as well, but I'm not sure if there's any reason to continue to using the URLs in the form of `https://repo.packagist.org/packages/drupal/composer_autoloader.json`. The packagist.org API docs have some explanation of their preferred way of looking up information.

**Comment by mscottford** [comment]
Just dropping one final note that `https://packagist.org/packages.json` follows the same rules as the ones I listed above.

**Comment by mscottford** [comment]
Another detail that I've discovered. The package details on packagist.org all include the `time` field, but it turns out that value is optional, and packages.drupal.org does appear to use it. Instead, you can grab a release timestamp from `extra.drupal.datestamp'. That appears to be the dates that are powering the release dates here: https://www.drupal.org/project/composer_autoloader/releases/7.x-1.0.

**Comment by mscottford** [comment]
But for asset-packagist.org, I'm not sure what do... I'm not seeing any dates in the data... https://asset-packagist.org/p/bower-asset/moment/ccb8b2d92b072498d01e6f22d7e4d84c74069432bf6e27ec1434cd4dbb551b8e.json

**Comment by SteveDesmond-ca** [comment]
Thanks for this bug report! As of v2.0.2, this at least now outputs a nicer message, and I've added a note in the README that only Packagist is currently supported.

**Comment by SteveDesmond-ca** [comment]
As it looks like you already discovered in [@corgibytes/freshli-lib](https://github.com/corgibytes/freshli-lib/blob/main/Corgibytes.Freshli.Lib/Languages/Php/ComposerRepository.cs#L87-L92), the reason this was failing is because the Drupal repository does not use the standard `time` field, but rather it's own custom `extras.drupal.datestamp`.

This case is now handled in the just-released v2.1.0 -- thanks again for the detailed report!
